### PR TITLE
Single storage of track angular fluxes for cases with vacuum boundary conditions only

### PR DESCRIPTION
**Pull request by GiudGiud**
This PR introduces the ONLYVACUUMBC macro, set at compilation time. 

Basically if the case only contains vacuum boundary conditions, it allows to store each track angular flux only once, thus saving a factor of two on memory on large cases, for which memory costs are clearly dominated by track angular flux storage.

Without domain decomposition, it's trivial
With domain decomposition, we have to be very careful during flux transfer time not to overwrite a flux that has not been sent yet. If an overwrite is about to happen, the flux is saved in the send_buffers to be transferred at the next round. This pre-filling of buffers can overflow them, when that happens, the buffer is doubled in size dynamically and a log is printed.

Communication costs overall go up, probably because the routine is now more complex and less optimized. 
Sweeping time remains dominant in the cases that are memory bound, for which we need something like this. 

Other changes :
- the code doesn't crash anymore when zero volume regions are used
- info logs now print the rank as well

### CODE REVIEW

**Comment by GiudGiud** [comment]
@mit-crpg/openmoc-dev anyone to review this ? It's quite critical for 3D full core calculations on less than 100 nodes (at 128 gb RAM) to reduce the storage of boundary angular fluxes.

**Comment by GiudGiud** [comment]
Merging with self-review Monday

File: src/CPUSolver.cpp [hunk 1]
```diff
@@ -490,27 +504,43 @@ void CPUSolver::setupMPIBuffers() {
...
     }
 
-    /* Setup Track communication information for all neighbor domains */
+    /* Estimate and print size of flux transfer buffers */
+    int size_mb = 2 * message_length * _send_buffers.size() * sizeof(float);
+    int max_size_mb;
+    MPI_Allreduce(&size_mb, &max_size_mb, 1, MPI_INT, MPI_MAX,
+                  _geometry->getMPICart());
+    log_printf(NORMAL, "Max track fluxes transfer buffer storage = %.2f MB",
+               max_size_mb / 1e6);
```
Comment (wuwenbin2006) [inline src/CPUSolver.cpp]: line 519~525 should be put after line 535. otherwise, size_mb is zero

File: src/CPUSolver.cpp [hunk 2]
```diff
@@ -663,6 +717,35 @@ void CPUSolver::deleteMPIBuffers() {
...
+    for (long i=0; i< _tracks_from_vacuum.size(); i++) {
+
+      long t_id = _tracks_from_vacuum.at(i) / 2;
+      int dir = _tracks_from_vacuum.at(i) - 2 * t_id;
+      for (int pe=0; pe<_fluxes_per_track; pe++)
+        _boundary_flux(t_id, dir, pe) = 0.;
+    }
+  }
+  else {
+    for (long t=0; t < _tot_num_tracks; t++)
```
Comment (wuwenbin2006) [inline src/CPUSolver.cpp]: OMP could be used here.

File: src/CPUSolver.cpp [hunk 3]
```diff
@@ -821,8 +905,10 @@ void CPUSolver::packBuffers(std::vector<long> &packing_indexes) {
   for (int i=0; i < num_domains; i++) {
     int send_domain = _neighbor_domains.at(i);
```
Comment (wuwenbin2006) [inline src/CPUSolver.cpp]: Seems send_domain is used nowhere.

File: src/CPUSolver.cpp [hunk 4]
```diff
@@ -821,8 +905,10 @@ void CPUSolver::packBuffers(std::vector<long> &packing_indexes) {
...
 
-    /* Reset send buffers */
-    int start_idx = _fluxes_per_track + 1;
+    /* Reset send buffers : start at beginning if the buffer has not been
+       prefilled, else start after what has been prefilled */
+    int start_idx = (_send_buffers_index.at(i)) * (_track_message_size) +
```
Comment (wuwenbin2006) [inline src/CPUSolver.cpp]: two pairs of brackets are useless here. remove them.

File: src/CPUSolver.cpp [hunk 5]
```diff
@@ -907,8 +1025,51 @@ void CPUSolver::transferAllInterfaceFluxes() {
...
+    /* Check if any rank needs to send buffers : because of the pre-filling
+       some nodes might have sent all their fluxes before others */
+    int need_to_send = 0;
+    for (int i=0; i < num_domains; i++) {
+
+      long* first_track_idx =
+        reinterpret_cast<long*>(&_send_buffers.at(i)[_fluxes_per_track+1]);
+      long first_track = first_track_idx[0];
+      if (first_track != -1)
+        need_to_send = 1;
```
Comment (wuwenbin2006) [inline src/CPUSolver.cpp]: A break of the for loop could be put here avoiding extra computation.

File: src/CPUSolver.cpp (hunk 1 above)
Comment (GiudGiud) [reply src/CPUSolver.cpp]: Good catch. Not sure why it wasnt printing 0 for me

File: src/CPUSolver.cpp (hunk 2 above)
Comment (GiudGiud) [reply src/CPUSolver.cpp]: Agreed

File: src/CPUSolver.cpp (hunk 3 above)
Comment (GiudGiud) [reply src/CPUSolver.cpp]: Good catch

File: src/CPUSolver.cpp (hunk 4 above)
Comment (GiudGiud) [reply src/CPUSolver.cpp]: Done

File: src/CPUSolver.cpp (hunk 5 above)
Comment (GiudGiud) [reply src/CPUSolver.cpp]: There's not much work done in that loop and not many iterations, it's fine to keep going

**Comment by GiudGiud** [comment]
@wuwenbin2006 thanks a lot for the review ! I think I addressed everything, and added an optimization for the track buffer size.  
In my current timings, angular flux exchange is 10% of the runtime for a full core case when using this special case of only-vacuum-boundary-conditions, so I would hold off on optimizing it for now. 

I agree the code is a bit clunky right now. We have to figure out a clean way to move some part of the solvers to another file. MPI communication routines are a good candidate. Once that's done, we can implement different ways of communicating across domains. Like I would like us to try a RMA implementation
Neighbor collective operations could be a good idea. Also, we have a cartesian mesh and if we find a MPI function that handles diagonal between domains as well we could get a pretty efficient implementation.

I think @friedmud 's PhD thesis will provide a lot of insight on how we can do angular flux exchange.

**Pull request merged**